marlowe/src/Language/Marlowe/Core/V1/Semantics/Next.hs

```diff
@@ -0,0 +1,175 @@
+--
+-- Module      :  $Headers
+-- License     :  Apache 2.0
+--
+-- Stability   :  Experimental
+-- Portability :  Portable
+--
+-- | Types for Marlowe semantics
+--
+-----------------------------------------------------------------------------
+
+
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+
+
+module Language.Marlowe.Core.V1.Semantics.Next
+  ( AmbiguousIntervalProvided(..)
+  , CaseIndex(..)
+  , Next(..)
+  , NextGeneralizedInput(..)
+  , next
+  ) where
+
+import Control.Applicative
+import Control.Monad.Cont ((=<<))
+import Data.Aeson
+import Data.Aeson.Types ()
+import Data.Bifunctor (Bifunctor(first))
+import Data.List.Index (indexed)
+import Data.List.NonEmpty as NEL (NonEmpty(..), nonEmpty, toList)
+import Data.Maybe (catMaybes)
+import Data.Text (pack)
+import Deriving.Aeson (Generic)
+import Language.Marlowe.Core.V1.Semantics
+import Language.Marlowe.Core.V1.Semantics.Types hiding (environment)
+import Language.Marlowe.Pretty (Pretty(..))
+import PlutusTx.Prelude (Bool(..), Either(..), Integer, Maybe(..), maybe, uncurry, ($), (.))
+import qualified Prelude as Haskell
+
+
+
+
+data Next = CanAdvance [NextGeneralizedInput]
+          | CanNotAdvance
+    deriving stock (Haskell.Show)
+
+data AmbiguousIntervalProvided = AmbiguousIntervalProvided
+
+newtype CaseIndex = CaseIndex {unIndex ::Integer}
+    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+    deriving anyclass (Pretty)
+    deriving (FromJSON,ToJSON)
+
+-- | Next Ranges of Inputs that can be applied to a given contract
+data NextGeneralizedInput
+  = CanDeposit CaseIndex Party AccountId Token Integer -- 1 possibility > 1 action
+  | CanChoose  CaseIndex ChoiceId [Bound] -- 1 possibility > many different actions
+  | CanNotify  CaseIndex -- 1 possibility > 1 action
+  deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+  deriving anyclass (Pretty)
+
+
+next :: Environment ->  State  ->  Contract -> Either AmbiguousIntervalProvided Next
+next  _ _ Close         = Right . CanAdvance $ []
+next  _ _ (When [] _ _) = Right CanNotAdvance   -- Contract Suspended in a time out
+next environment state (When (x:xs) _ _)
+  = Right
+    . maybe
+          CanNotAdvance -- Contract Suspended (only falsified Notifies evaluated in the When)
+        ( CanAdvance . keepFirstNotifyTrueAndRemoveFollowings  )
+    . nextGeneralizedInputsMaybe environment state
+    $ x :| xs
+next environment state  contract  = uncurry (next environment) =<< reduceContract environment state contract
+
+
+keepFirstNotifyTrueAndRemoveFollowings ::  NonEmpty NextGeneralizedInput -> [NextGeneralizedInput]
```

Comment:
```suggestion
keepFirstNotifyTrueAndRemoveFollowings ::  NonEmpty NextGeneralizedInput -> [NextGeneralizedInput]
```
Why consider empty list as a special case and then we have a `NonEmpty` list, if we could just have a normal list that considers both cases and is simpler and more widely understood?

In other words, instead of:
```haskell
  CanApplyInputs (NonEmpty NextGeneralizedInput)
| Suspended
``` 
Why not just write `[NextGeneralizedInput]`? But because we need to consider apply with no inputs, we actually need something like `(Bool, [NextGeneralizedInput])`, or something along those lines